### PR TITLE
Create generic AlterTableTask that can replace ADD/DROP/RENAME COLUMN tasks

### PR DESCRIPTION
**Pull request by jeeminso**
## Summary of the changes / Why this improves CrateDB
originate from https://github.com/crate/crate/pull/15004#discussion_r1416796459.

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Review by mfussenegger — COMMENTED** [review]
What's the motivation for this? It's very much intentional that the supplier is not generally available, but only used when writing to the cluster state.

The intention is to help ensure it's derived from the ClusterState Builder, and that it's used in an atomic cluster state update.

**Comment by jeeminso** [comment]
> What's the motivation for this? It's very much intentional that the supplier is not generally available, but only used when writing to the cluster state.
> 
> The intention is to help ensure it's derived from the ClusterState Builder, and that it's used in an atomic cluster state update.

Thank you for the explanation. This PR is for https://github.com/crate/crate/pull/15004#discussion_r1416796459. I was experimenting if I could remove `OidSupplier` parameter from `DocTableInfo.addColumn` which is not needed for `dropColumn` or `renameColumn` thinking `OidSupplier` per table makes sense. I will revert and move on.

**Comment by jeeminso** [comment]
Could you please have a quick look at this draft @mfussenegger ? If this seems ok, I will port existing `...ColumnTaskTests` to `DocTableInfoTest`.

File: server/src/main/java/io/crate/execution/ddl/tables/TableModifyTask.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+        newTable.writeTo(createMapperService, metadata, metadataBuilder);
+        Metadata newMetadata = metadataBuilder.build();
+        // Ensure new table can still be parsed
+        docTableInfoFactory.create(relationName, newMetadata);
+        return ClusterState.builder(currentState)
+            .metadata(newMetadata)
+            .build();
+    }
+
+    public interface TableModifyExecutor<T extends AcknowledgedRequest<?>> extends TriFunction<T, DocTableInfo, Metadata.Builder, DocTableInfo> {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/execution/ddl/tables/TableModifyTask.java:76]:
```suggestion
    public interface TableModifyExecutor<T> extends TriFunction<T, DocTableInfo, Metadata.Builder, DocTableInfo> {
```

The additional constraint is not used

**Review by mfussenegger — COMMENTED** [review]
Looks good to me.

I'd probably change the names slightly:

`TableModifyTask -> AlterTableTask` - to stick with existing terminology
`TableModifyExecutor -> AlterTableOperator` or `UpdateDocTableInfo` (and depending on that, also update the `executor` field - it's misleading - it's not an executor but the update operation.

File: server/src/main/java/io/crate/execution/ddl/tables/TableModifyTask.java (hunk 1 above)
Comment (jeeminso) [reply server/src/main/java/io/crate/execution/ddl/tables/TableModifyTask.java]: Thank you for reviewing, please check the commit for updates.

File: server/src/main/java/io/crate/metadata/doc/DocTableInfo.java [hunk 2]
```diff
@@ -216,7 +217,8 @@ public DocTableInfo(RelationName ident,
                         Version versionCreated,
                         @Nullable Version versionUpgraded,
                         boolean closed,
-                        Set<Operation> supportedOperations) {
+                        Set<Operation> supportedOperations,
+                        NodeContext nodeContext) {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/metadata/doc/DocTableInfo.java:221]: Can you revert this part?

The `NodeContext` is not a part of a table and doesn't belong here.
 
Either extend the `AlterTableOperator` interface with another parameter (and it doesn't need to extend a Function interface if that get's in the way, we don't make use of that anyway)

Or remove the static constant and pass it along when needed.

**Pull request merged**